ratelimits/limit.go

```diff
@@ -0,0 +1,137 @@
+package ratelimits
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"strings"
+
+	"github.com/letsencrypt/boulder/config"
+	"github.com/letsencrypt/boulder/strictyaml"
+)
+
+type limit struct {
+	// Burst specifies maximum concurrent allowed requests at any given time. It
+	// must be greater than zero.
+	Burst int64
+
+	// Count is the number of requests allowed per period. It must be greater
+	// than zero.
+	Count int64
+
+	// Period is the duration of time in which the count (of requests) is
+	// allowed. It must be greater than zero.
+	Period config.Duration
+}
+
+type limits map[string]limit
+
+func parseDefaultName(k string) (string, error) {
+	name, ok := stringToName[k]
+	if !ok {
+		return "", fmt.Errorf(
+			"unrecognized limit %q, must be one in %q", k, limitNames)
+	}
+	return nameToIntString(name), nil
+}
+
+func parseOverrideNameId(k string) (string, string, error) {
+	nameAndId := strings.SplitN(k, ":", 2)
+	nameStr := nameAndId[0]
+	if nameStr == "" {
+		return "", "", fmt.Errorf("empty name in override %q, must be 'name:id'", k)
+	}
+	id := nameAndId[1]
+	if id == "" {
+		return "", "", fmt.Errorf("empty id in override %q, must be 'name:id'", k)
+	}
+
+	name, ok := stringToName[nameStr]
+	if !ok {
+		return "", "", fmt.Errorf(
+			"unrecognized limit %q, must be one in %q", nameStr, limitNames)
+	}
+
+	if idIsIPv4Addr(name) {
+		ip := net.ParseIP(id)
+		if ip == nil || ip.To4() == nil {
+			return "", "", fmt.Errorf(
+				"invalid addr %q, in override %q, must be IPv4", id, k)
+		}
+	}
+
+	if idIsIPv6Range(name) {
+		_, net, err := net.ParseCIDR(id)
+		if err != nil {
+			return "", "", fmt.Errorf(
+				"invalid id %q, in override %q, must be IPv6 CIDR range", id, k)
+		}
+		if net.IP.To4() != nil {
+			return "", "", fmt.Errorf(
+				"invalid CIDR %q, in override %q, must be IPv6 CIDR range", id, k)
+		}
+		ones, _ := net.Mask.Size()
+		if ones != 48 {
+			return "", "", fmt.Errorf(
+				"invalid range %q, in override %q, must be /48,", id, k)
+		}
+	}
```

Comment:
I'm confused: under what circumstances would the `name` be an IPv4 address or IPv6 range? Isn't the `name` always supposed to be the name of one of our rate limits, which we explicitly enumerate?

Reply:
We discussed this offline. I'll post the reworked version of this that I proposed.